File: Scala/1-callback.scala
object Callback {

  def func(data: Option[String], callback: (Either[Exception, String]) => Int): Int =
    data map {d =>
      callback(Right(s"Data: $d"))
    } getOrElse {
      callback(Left(new Exception("Exception occurred!")))
    }

  def simpleCallback(input: Either[Exception, String]): Int =
    input map { s: String =>
      println(s)
      1 // ok, status 1
    } getOrElse {
      println(s"Data is not present!\n${input.left.get.getMessage}")
      -1 // NOT ok, status -1
    }

  def main(args: Array[String]): Unit = {
    println(s"None status = ${func(None, simpleCallback)}")

    println(s"Some status = ${func(Some("123"), simpleCallback)}")
  }
}

File: Python/1-callback.py
def fn(par, callback):
    if not par:
        callback(Exception("Parametr needed"))
        return
    callback(None, f'Data: {par}')
    return 'Value'

def callback(err, data=None):
    if err:
        raise err
    print(data)

res = fn('example', callback)
fn(None, callback)
print(res)


